File: android/app/src/main/java/com/poshanforlife/android/feature/practitioner/leads/LeadsViewModel.kt
package com.poshanforlife.android.feature.practitioner.leads

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.poshanforlife.android.core.data.LeadRepository
import com.poshanforlife.android.core.network.LeadDetailDto
import com.poshanforlife.android.core.network.LeadListItemDto
import com.poshanforlife.android.core.network.LeadSummaryDto
import com.poshanforlife.android.core.network.Result
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

private const val SEARCH_DEBOUNCE_MS = 300L

sealed class LeadListUiState {
    data object Loading : LeadListUiState()
    data class Success(val leads: List<LeadListItemDto>, val summary: LeadSummaryDto?) : LeadListUiState()
    data class Error(val message: String) : LeadListUiState()
}

sealed class LeadDetailUiState {
    data object Loading : LeadDetailUiState()
    data class Success(val lead: LeadDetailDto) : LeadDetailUiState()
    data class Error(val message: String) : LeadDetailUiState()
}

sealed class LeadActionState {
    data object Idle : LeadActionState()
    data object InFlight : LeadActionState()
    data class Error(val message: String) : LeadActionState()
}

/**
 * Backs LeadListScreen and LeadDetailScreen (a per-screen Hilt instance
 * each, same dual-purpose convention as DocumentsViewModel/
 * PatientManagementViewModel) — which half of the state a screen reads
 * depends on whether "leadId" is present in its route. Reused verbatim by
 * both the practitioner and admin nav graphs, same as Documents/Orders.
 */
@HiltViewModel
class LeadsViewModel @Inject constructor(
    private val leadRepository: LeadRepository,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {

    private val leadId: String? = savedStateHandle["leadId"]

    // ---- list ----------------------------------------------------------

    private val _listState = MutableStateFlow<LeadListUiState>(LeadListUiState.Loading)
    val listState: StateFlow<LeadListUiState> = _listState.asStateFlow()

    private val _stageFilter = MutableStateFlow<String?>(null)
    val stageFilter: StateFlow<String?> = _stageFilter.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    private var listJob: Job? = null

    // ---- detail ----------------------------------------------------------

    private val _detailState = MutableStateFlow<LeadDetailUiState>(LeadDetailUiState.Loading)
    val detailState: StateFlow<LeadDetailUiState> = _detailState.asStateFlow()

    private val _stageChangeState = MutableStateFlow<LeadActionState>(LeadActionState.Idle)
    val stageChangeState: StateFlow<LeadActionState> = _stageChangeState.asStateFlow()

    private val _activityState = MutableStateFlow<LeadActionState>(LeadActionState.Idle)
    val activityState: StateFlow<LeadActionState> = _activityState.asStateFlow()

    private val _followupState = MutableStateFlow<LeadActionState>(LeadActionState.Idle)
    val followupState: StateFlow<LeadActionState> = _followupState.asStateFlow()

    init {
        if (leadId == null) refreshList() else loadDetail()
    }

    /** Immediate — used for pull-to-refresh and filter chip taps. */
    fun refreshList() {
        listJob?.cancel()
        listJob = viewModelScope.launch { performListLoad() }
    }

    fun onStageFilterChange(stage: String?) {
        _stageFilter.update { stage }
        refreshList()
    }

    /** Debounced + cancels any in-flight search, so a slow response to a stale keystroke can't clobber a newer one. */
    fun onSearchQueryChange(query: String) {
        _searchQuery.update { query }
        listJob?.cancel()
        listJob = viewModelScope.launch {
            delay(SEARCH_DEBOUNCE_MS)
            performListLoad()
        }
    }

    private suspend fun performListLoad() {
        _isRefreshing.update { true }
        when (val result = leadRepository.listPipeline(stage = _stageFilter.value, search = _searchQuery.value.ifBlank { null })) {
            is Result.Success -> _listState.update { LeadListUiState.Success(result.data.leads, result.data.summary) }
            is Result.Error -> _listState.update { LeadListUiState.Error(result.message) }
            Result.Loading -> Unit
        }
        _isRefreshing.update { false }
    }

    fun loadDetail() {
        val id = checkNotNull(leadId)
        viewModelScope.launch {
            _detailState.update { LeadDetailUiState.Loading }
            when (val result = leadRepository.getLead(id)) {
                is Result.Success -> _detailState.update { LeadDetailUiState.Success(result.data) }
                is Result.Error -> _detailState.update { LeadDetailUiState.Error(result.message) }
                Result.Loading -> Unit
            }
        }
    }

    fun changeStage(newStage: String) {
        val id = checkNotNull(leadId)
        viewModelScope.launch {
            _stageChangeState.update { LeadActionState.InFlight }
            when (val result = leadRepository.updateStage(id, newStage)) {
                is Result.Success -> {
                    _detailState.update { LeadDetailUiState.Success(result.data) }
                    _stageChangeState.update { LeadActionState.Idle }
                }
                is Result.Error -> _stageChangeState.update { LeadActionState.Error(result.message) }
                Result.Loading -> Unit
            }
        }
    }

    fun logActivity(activityType: String, description: String) {
        val id = checkNotNull(leadId)
        viewModelScope.launch {
            _activityState.update { LeadActionState.InFlight }
            when (val result = leadRepository.addActivity(id, activityType, description)) {
                is Result.Success -> {
                    _activityState.update { LeadActionState.Idle }
                    loadDetail()
                }
                is Result.Error -> _activityState.update { LeadActionState.Error(result.message) }
                Result.Loading -> Unit
            }
        }
    }

    fun scheduleFollowup(followupAt: String, message: String?) {
        val id = checkNotNull(leadId)
        viewModelScope.launch {
            _followupState.update { LeadActionState.InFlight }
            when (val result = leadRepository.scheduleFollowup(id, followupAt, message)) {
                is Result.Success -> {
                    _detailState.update { LeadDetailUiState.Success(result.data) }
                    _followupState.update { LeadActionState.Idle }
                }
                is Result.Error -> _followupState.update { LeadActionState.Error(result.message) }
                Result.Loading -> Unit
            }
        }
    }
}

File: ios/PoshanForLife/Features/Leads/LeadsViewModel.swift
import Foundation

/// Backs both `LeadListView` and `LeadDetailView` — one class, mode decided
/// by whether a `leadId` was given, same shape as `PatientManagementViewModel`.
@MainActor
final class LeadsViewModel: ObservableObject {

    enum ActionState: Equatable {
        case idle, inFlight
        case failure(String)
    }

    let leadId: String?

    // MARK: List

    @Published private(set) var listState: CardState<[LeadListItem]> = .loading
    @Published private(set) var summary: LeadSummary?
    @Published var searchQuery: String = "" {
        didSet { scheduleSearch() }
    }
    @Published var stageFilter: LeadStage? {
        didSet { Task { await performListLoad() } }
    }

    // MARK: Detail

    @Published private(set) var detailState: CardState<LeadDetail> = .loading
    @Published private(set) var stageChangeState: ActionState = .idle
    @Published private(set) var activityState: ActionState = .idle
    @Published private(set) var followupState: ActionState = .idle

    private let repository: LeadsRepository
    private var searchTask: Task<Void, Never>?

    init(leadId: String? = nil, repository: LeadsRepository) {
        self.leadId = leadId
        self.repository = repository
    }

    // MARK: - List

    func loadList() async {
        listState = .loading
        await performListLoad()
    }

    func refreshList() async {
        await performListLoad()
    }

    /// Debounced: cancels any in-flight search so a slow response to a stale
    /// keystroke can't land after — and overwrite — a newer one. Same 300ms
    /// as `PatientManagementViewModel`/Android's lead search.
    private func scheduleSearch() {
        searchTask?.cancel()
        searchTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 300_000_000)
            guard !Task.isCancelled else { return }
            await self?.performListLoad()
        }
    }

    private func performListLoad() async {
        switch await repository.list(stage: stageFilter, search: searchQuery) {
        case .success(let response):
            listState = .success(response.leads)
            summary = response.summary
        case .failure(let error):
            listState = .failure(error.message)
        }
    }

    // MARK: - Detail

    func loadDetail() async {
        guard let leadId else { return }
        detailState = .loading
        detailState = (await repository.detail(id: leadId)).cardState
    }

    /// Reloads the detail on success rather than patching `stage` in place —
    /// the backend auto-logs a `stage_change` activity, and the timeline
    /// should reflect it without a second round trip.
    func changeStage(to stage: LeadStage) async {
        guard let leadId else { return }
        stageChangeState = .inFlight
        switch await repository.updateStage(id: leadId, stage: stage) {
        case .success(let detail):
            detailState = .success(detail)
            stageChangeState = .idle
        case .failure(let error):
            stageChangeState = .failure(error.message)
        }
    }

    func logActivity(type: LeadActivityType, description: String) async -> Bool {
        guard let leadId else { return false }
        activityState = .inFlight
        switch await repository.addActivity(id: leadId, type: type, description: description) {
        case .success:
            activityState = .idle
            await loadDetail()
            return true
        case .failure(let error):
            activityState = .failure(error.message)
            return false
        }
    }

    func scheduleFollowup(at date: Date, message: String?) async -> Bool {
        guard let leadId else { return false }
        followupState = .inFlight
        switch await repository.scheduleFollowup(id: leadId, at: date, message: message) {
        case .success(let detail):
            detailState = .success(detail)
            followupState = .idle
            return true
        case .failure(let error):
            followupState = .failure(error.message)
            return false
        }
    }
}

private extension Result where Failure == APIError {
    var cardState: CardState<Success> {
        switch self {
        case .success(let value): return .success(value)
        case .failure(let error): return .failure(error.message)
        }
    }
}
